new README docs section (#576)
jonataslaw/getx: README.md

## Change Theme

> Please do not use any higher level widget than `GetMaterialApp` in order to update it. This can trigger duplicate keys. A lot of people are used to the prehistoric approach of creating a "ThemeProvider" widget just to change the theme of your app, and this is definitely NOT necessary with **GetX™**.

Reviewer: prehistoric? aren't this the usual way of doing it without GetX?

Reviewer: @jonataslaw

Reviewer 2: Just for my understanding, `Get` needs to be replaced with `GetX™` everywhere?
Because there are still some places in this PR where it is `Get`

Reviewer: @stefandevo the ideal is to replace everywhere

),
```

> ####ObxValue

Reviewer: ```suggestion
#### ObxValue
```

> Similar to [`ValueBuilder`](#valuebuilder), but this is the Reactive version, you pass a Rx instance (remember the magical .obs?) and 

Reviewer: Nice idea man, put links in between the docs helps a lot

),
```

## Useful tips

`.obs`ervables (also known as _Rx_ Types) have a wide variety of internal methods and operators.

> Is very common to _believe_ that a property with `.obs` **IS** the actual value... but make no mistake!
We avoid the Type declaration of the variable, because Dart's compiler is smart enough, and the code
looks cleaner, but:
```dart
var message = 'Hello world'.obs;
print( 'Message "$message" has Type ${message.runtimeType}');
```
Even if `message` _prints_ the actual String value, the Type is **RxString**!

So, you can't do `message.substring( 0, 4 )`.
You have to access the real `value` inside the _observable_: 
The most "used way" is `.value`, but, did you know that you can also use...

```dart
final name = 'GetX'.obs;
// only "updates" the stream, if the value is different from the current one. 
name.value = 'Hey';

// this weird (and kinda cool) assignment, updates the stream no matter what

> // it takes nulls, or same value... but rebuilds the observers.

Reviewer: oh so assigning like that will rebuild even if the value is the same? This kinda break one of the principles of reactive state manager: only rerender when necessary.
I think this operator should respect this principle too

name << 'Hey'; // !

// All Rx properties are "callable" and returns the new value.
// but this approach does not accepts `null`, the UI will not rebuild.
name('Hello');

// is like a getter, prints 'Hello'.
name() ;

/// numbers:

final count = 0.obs; 

// you can just most basic operators acts on the property!
count + 1;
 
// Watch out! this is only valid if `count` is not final, but var
count += 1;

// You can also compare against values:
count > 2;

/// booleans:

final flag = false.obs;

// switches the value between true/false
flag.toggle();

/// all types:

// Sets the `value` to null.

> flag.nil();

Reviewer: although i can understand that nil it null, this is a word that does not exist in dart, and people that never had contact with a language that used this may be confused.

The ideal in my opinion is `flag.setNull()`

i know this is not the PR to say this but i only thought about that now 😅

Author: I like short methods names haha. And having LISP nil had a pun intended with flutter. But anyway, open to suggestions, whatever the name is a must to have such utility to initialize with null

final count=0.obs..setNull();

// looks much cleaner no?
final bla=0.obs..nil();

Reviewer: i don't get the joke 😅 

well, i don't think it cleaner to be honest 😅 , but that is only my opinion

// All toString(), toJson() operations are passed down to the `value`
print( count ); // calls `toString()` inside  for RxInt

final abc = [0,1,2].obs;
// Converts the value to a json Array, prints RxList
// Json is supported by all Rx types!
print('json: ${jsonEncode(abc)}, type: ${abc.runtimeType}'); 

// RxMap, RxList and RxSet are special Rx types, that extends their native types.
// but you can work with a List as a regular list, although is reactive!
abc.add(12); // pushes 12 to the list, and UPDATES the stream.
abc[3]; // like Lists, reads the index 3.

// equality works with the Rx and the value, but hashCode is always taken from the value
final number = 12.obs;
print( number == 12 ); // prints > true

/// Custom Rx Models:

> // toJson(), toString() are deffered to the child, so you can implement override them, and print() the observable directly. 

Reviewer: ```suggestion
// toJson(), toString() are deferred to the child, so you can implement override on them, and print() the observable directly. 
```

#### GetxService

This class is like a `GetxController`, it shares the same lifecycle ( `onInit()`, `onReady()`, `onClose()`).
But has no "logic" inside of it. It just notifies **GetX** Dependency Injection system, that this subclass
**can not** be removed from memory. 

So is super useful to keep your "Services" always reachable and active with `Get.find()`. Like:
`ApiService`, `StorageService`, `CacheService`. 

```dart
Future<void> main() async {
  await initServices(); /// AWAIT SERVICES INITIALIZATION.
  runApp(SomeApp());
}

/// Is a smart move to make your Services intiialize before you run the Flutter app.
/// as you can control the execution flow (maybe you need to load some Theme configuration, 
/// apiKey, language defined by the User... so load SettingService before running ApiService.
/// so GetMaterialApp() doesnt have to rebuild, and takes the values directly.    
void initServices() async {
  print('starting services ...');
  /// Here is where you put get_storage, hive, shared_pref initialization.
  /// or moor connection, or whatever that's async.
  await Get.putAsync(() => DbService().init());

>   await Get.put(SettingsService()).init();

Reviewer: ```suggestion
  await Get.putAsync(SettingsService()).init();
```

  print('All services started...');
}

class DbService extends GetxService {
  Future<DbService> init() async {
    print('$runtimeType delays 2 sec');
    await 2.delay();
    print('$runtimeType ready!');
    return this;
  }
}

class SettingsService extends GetxService {
  void init() async {

>     print('$runtimeType delays 2 sec');

Reviewer: ```suggestion
    print('$runtimeType delays 1 sec');
```

    await 1.delay();
    print('$runtimeType ready!');
  }
}

```

The only way to actually delete a `GetxService`, is with `Get.reset()` which is like a
"Hot Reboot" of your app. So remember, if you need absolute persistance of a class instance during the
lifetime of your app, use `GetxService`. 
  

 

> 

Reviewer: the markdown ignores by default more than one line, but even so, it's a good practice to not let more than 1 blank line
